cocotb/_build_libs.py

```diff
@@ -0,0 +1,438 @@
+            src_filename = os.path.join(self.build_lib, filename)
+
+            os.makedirs(os.path.dirname(dest_filename), exist_ok=True)
+
+            copy_file(
+                src_filename, dest_filename, verbose=self.verbose, dry_run=self.dry_run
+            )
+            if ext._needs_stub:
+                self.write_stub(package_dir or os.curdir, ext, True)
+
+
+# Allow to load dependencies from direcotry where vpi/vhpi/fli library is located on osx
+def _extra_link_args(lib_name):
+    if sys.platform == "darwin":
+        return ["-Wl,-install_name,@loader_path/%s.so" % lib_name]
+    else:
+        return []
+
+
+def _get_python_lib_link():
+    """ Get name of python library used for linking """
+
+    if sys.platform == "darwin":
+        ld_library = sysconfig.get_config_var("LIBRARY")
+    else:
+        ld_library = sysconfig.get_config_var("LDLIBRARY")
+
+    if ld_library is not None:
+        python_lib_link = os.path.splitext(ld_library)[0][3:]
+    else:
+        python_version = sysconfig.get_python_version().replace(".", "")
+        python_lib_link = "python" + python_version
+
+    return python_lib_link
+
+
+def _get_python_lib():
+    """ Get the library for embedded the python interpreter """
+
+    if os.name == "nt":
+        python_lib = _get_python_lib_link() + "." + _get_lib_ext_name()
+    else:
+        python_lib = "lib" + _get_python_lib_link() + "." + _get_lib_ext_name()
+
+    return python_lib
+
+
+_base_warns = ["-Wall", "-Wextra", "-Wconversion", "-Wcast-qual", "-Wwrite-strings"]
+_cc_warns = _base_warns + ["-Wstrict-prototypes", "-Waggregate-return"]
+_ccx_warns = _base_warns + ["-Wnon-virtual-dtor", "-Woverloaded-virtual"]
+
+# TODO Enable warnings when fixed
+_extra_cc_compile_args = ["-std=gnu99"] + _cc_warns
+_extra_cxx_compile_args = ["-std=c++11"] + _ccx_warns
+
+# We need to build common libraries for each simulator to avoid issues with 
+# loading libraries from different directories in multi os configuration and 
+# cross dependencies without setting globally LD_LIBRARY_PATH or similar. 
+# `LIB_DIR` is used to point to this directory.
+def _get_common_lib_ext(include_dir, share_lib_dir, sim_define):
+
+    #
+    #  libcocotbutils
+    #
+    libcocotbutils = Extension(
+        os.path.join("cocotb", "libs", sim_define.lower(), "libcocotbutils"),
+        include_dirs=[include_dir],
+        sources=[os.path.join(share_lib_dir, "utils", "cocotb_utils.c")],
+        extra_link_args=_extra_link_args("libcocotbutils"),
+        extra_compile_args=_extra_cc_compile_args,
+    )
+
+    #
+    #  libgpilog
+    #
+    gpilog_library_dirs = []
+    if sys.platform == "darwin":
+        gpilog_library_dirs = [sysconfig.get_config_var("LIBDIR")]
+
+    libgpilog = Extension(
+        os.path.join("cocotb", "libs", sim_define.lower(), "libgpilog"),
+        include_dirs=[include_dir],
+        libraries=[_get_python_lib_link(), "pthread", "m", "cocotbutils"],
```

Comment:
Is this needed? I thought extension modules were automatically linked against python
```suggestion
        libraries=["pthread", "m", "cocotbutils"],
```

Reply:
One gets
```
gpivpi:`/cocotb/libs/icarus/gpivpi.vpl' failed to open using dlopen() because:
 /cocotb/libs/icarus/libcocotb.so: undefined symbol: _Py_NoneStruct.
```

Reply:
sometimes: https://travis-ci.org/themperek/cocotb/builds/643871989 (very interesting)

Reply:
Ha: `Changed in version 3.8: On Unix, C extensions are no longer linked to libpython except on Android and Cygwin.` here: https://docs.python.org/3/distutils/apiref.html#distutils.core.Extension

Reply:
It sounds like `_get_python_lib_link` should be in `libcocotb`, since that's the one that actually embeds python?

Reply:
Can be just followed the Makefiles https://github.com/cocotb/cocotb/blob/master/cocotb/share/lib/gpi_log/Makefile

Will check.

Reply:
The makefile has `_get_python_lib_link` for `cocotb` too, I think.

Reply:
Does this have to be done like this (is this needed)?https://github.com/cocotb/cocotb/blob/master/cocotb/share/lib/embed/gpi_embed.c#L138

It actully looks as not need if we link correctly?  (I am definitely not an expert here)

Reply:
Not an expert either, but that sound quite plausible. Let's not remove that in this PR though.

Reply:
What was the verdict here?

Reply:
Do like it is in makefiles (`_get_python_lib_link()` to `libcocotb` and `libgpilog` for now).

Reply:
Alright, that seems fair for now.

On that note thought - the makefiles don't seem to add `libm` and `libpthread` to `gpilog` - is the difference intentional?

Reply:
Actully is used in `simulator` and `libgpilog` will correct.

Reply:
My question here about `m` and `pthread` still applies - I don't see them in the old makefile - can you point me to what I'm missing?

Reply:
Searching in the repo for `pthread` finds no interesting results: https://github.com/cocotb/cocotb/search?q=pthread&unscoped_q=pthread

Reply:
sorry missed this will check

Reply:
Success, I assume?

Reply:
Yes. You have a good eye ;-)